src/python/pants/bin/remote_pants_runner.py

```diff
@@ -86,6 +89,44 @@ def _setup_logging(self):
     root.setLevel(log_level)
     root.addHandler(handler)
 
+  @staticmethod
+  def _backoff(attempt):
+    """Minimal backoff strategy for daemon restarts."""
+    time.sleep(attempt + (attempt - 1))
+
+  def _run_pants_with_retry(self, port, retries=3):
+    """Runs pants remotely with retry and recovery for nascent executions."""
+    attempt = 1
+    while 1:
+      logger.debug(
+        'connecting to pantsd on port {} (attempt {}/{})'.format(port, attempt, retries)
+      )
+      try:
+        return self._connect_and_execute(port)
+      except self.RECOVERABLE_EXCEPTIONS as e:
+        if attempt > retries:
+          raise self.Fallback(e)
+
+        self._backoff(attempt)
+        logger.warn(
+          'pantsd was unresponsive on port {}, retrying ({}/{})'
+          .format(port, attempt, retries)
+        )
+
+        # One possible cause of the daemon being non-responsive during an attempt might be if a
+        # another lifecycle operation is happening concurrently (incl teardown). To account for
+        # this, we won't begin attempting restarts until at least 1 second has passed (1 attempt).
+        if attempt > 1:
+          port = self._restart_pantsd()
```

Comment:
Is it possible there are dangling pantd process that needs to be killed before restarting one?

Reply:
`PantsDaemon.launch()` already calls `ProcessManager.terminate()` to handle that case here: https://github.com/pantsbuild/pants/blob/master/src/python/pants/pantsd/pants_daemon.py#L393

Reply:
Ah ok. Thank you!